src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -418,117 +255,56 @@ private Type EmitCoercionCode(AssociativeNode arg, Type argType, Type param)
             }
             else
             {
-                localBuilder = DeclareLocal(typeof(Source[]), "array to coerce");
+                var localBuilder = DeclareLocal(typeof(IEnumerable<Source>), "collection to coerce");
                 currentVarIndex = localBuilder.LocalIndex;
 
                 EmitOpCode(OpCodes.Stloc, currentVarIndex);
                 EmitOpCode(OpCodes.Ldloc, currentVarIndex);
             }
-            // Find length for array to be coerced (already on top of eval stack), len
-            EmitOpCode(OpCodes.Ldlen);
-            EmitOpCode(OpCodes.Conv_I4);
-
-            // var newarr = new Target[len];
-            EmitOpCode(OpCodes.Newarr, typeof(Target));
-
-            // Declare new array to store coerced values
-            var t = typeof(Target[]);
-            localBuilder = DeclareLocal(t, "coerced array");
-            var newArrIndex = localBuilder.LocalIndex;
-            EmitOpCode(OpCodes.Stloc, newArrIndex);
-
-            // Emit for loop to loop over array and convert.
-
-            // i = 0;
-            localBuilder = DeclareLocal(typeof(int), "for loop counter");
-            var counterIndex = localBuilder.LocalIndex;
-            EmitOpCode(OpCodes.Ldc_I4_0);
-            EmitOpCode(OpCodes.Stloc, counterIndex);
-
-            var loopBodyLabel = DefineLabel();
-            var loopCondLabel = DefineLabel();
-
-            EmitOpCode(OpCodes.Br_S, loopCondLabel.Value);
 
-            // newarr[i] = (Target)arr[i];
-            MarkLabel(loopBodyLabel.Value, "label:body");
-
-            EmitOpCode(OpCodes.Ldloc, newArrIndex);
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-
-            
-            EmitOpCode(OpCodes.Ldloc, currentVarIndex);
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-
-            if (typeof(Source) == typeof(double))
-            {
-                EmitOpCode(OpCodes.Ldelem_R8);
-            }
-            else if (typeof(Source) == typeof(long))
-            {
-                EmitOpCode(OpCodes.Ldelem_I8);
-            }
-            else if (typeof(Source) == typeof(int))
+            EmitOpCode(OpCodes.Ldnull);
+            MethodInfo convertMethod = null;
+            if (typeof(Target) == typeof(int))
             {
-                EmitOpCode(OpCodes.Ldelem_I4);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToInt32), new[] { typeof(Source) });
```

Comment:
perhaps to make compilation faster we could cache these method lookups at compiler startup time instead of looking them up each time we go through this function.